[SPARK-30482][SQL][CORE][TESTS] Add sub-class of `AppenderSkeleton` reusable in tests

MaxGekk opened the pull request:
### What changes were proposed in this pull request?
In the PR, I propose to define a sub-class of `AppenderSkeleton` in `SparkFunSuite` and reuse it from other tests. The class stores incoming `LoggingEvent` in an array which is available to tests for future analysis of logged events.

### Why are the changes needed?
This eliminates code duplication in tests.

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
By existing test suites - `CSVSuite`, `OptimizerLoggingSuite`, `JoinHintSuite`, `CodeGenerationSuite` and `SQLConfSuite`.

MaxGekk committed:
Move LogAppender to SparkFunSuite

MaxGekk committed:
Reuse LogAppender from SparkFunSuite

MaxGekk committed:
Dedup CSVSuite

MaxGekk committed:
Dedup OptimizerLoggingSuite

MaxGekk committed:
Dedup JoinHintSuite

MaxGekk committed:
Dedup CodeGenerationSuite

MaxGekk committed:
Merge remote-tracking branch 'remotes/origin/master' into dedup-appender-skeleton

MaxGekk committed:
Dedup SQLConfSuite

MaxGekk commented:
@maropu Please, review the PR.

SparkQA commented:
**[Test build #116478 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116478/testReport)** for PR 27166 at commit [`088c414`](https://github.com/apache/spark/commit/088c4140734632d283d46fafb175b349486af2c8).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21265/
Test PASSed.

MaxGekk commented:
It follows up https://github.com/apache/spark/pull/27092#discussion_r364018890

SparkQA commented:
**[Test build #116478 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116478/testReport)** for PR 27166 at commit [`088c414`](https://github.com/apache/spark/commit/088c4140734632d283d46fafb175b349486af2c8).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116478/
Test FAILed.

MaxGekk commented:
jenkins, retest this, please

SparkQA commented:
**[Test build #116484 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116484/testReport)** for PR 27166 at commit [`088c414`](https://github.com/apache/spark/commit/088c4140734632d283d46fafb175b349486af2c8).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21271/
Test PASSed.

maropu reviewed (approved):
Looks fine if the tests passed, thanks, @MaxGekk ! cc: @dongjoon-hyun @HyukjinKwon

SparkQA commented:
**[Test build #116484 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116484/testReport)** for PR 27166 at commit [`088c414`](https://github.com/apache/spark/commit/088c4140734632d283d46fafb175b349486af2c8).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116484/
Test FAILed.

HyukjinKwon commented:
retest this please

SparkQA commented:
**[Test build #116489 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116489/testReport)** for PR 27166 at commit [`088c414`](https://github.com/apache/spark/commit/088c4140734632d283d46fafb175b349486af2c8).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21276/
Test PASSed.

SparkQA commented:
**[Test build #116489 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116489/testReport)** for PR 27166 at commit [`088c414`](https://github.com/apache/spark/commit/088c4140734632d283d46fafb175b349486af2c8).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116489/
Test FAILed.

MaxGekk commented:
I don't think the failure of python tests is related to the PR.

MaxGekk committed:
Merge remote-tracking branch 'origin/master' into dedup-appender-skeleton

SparkQA commented:
**[Test build #116504 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116504/testReport)** for PR 27166 at commit [`6401164`](https://github.com/apache/spark/commit/640116443b362fd130f1da32c7e57365831e33fa).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21291/
Test PASSed.

SparkQA commented:
**[Test build #116504 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116504/testReport)** for PR 27166 at commit [`6401164`](https://github.com/apache/spark/commit/640116443b362fd130f1da32c7e57365831e33fa).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class FilterStatistics(`
  * `class EventLogFileCompactor(`
  * `case class CompactionResult(code: CompactionResultCode.Value, compactIndex: Option[Long])`
  * `case class AlterNamespaceSetOwner(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116504/
Test FAILed.

maropu commented:
The related thread: https://github.com/apache/spark/pull/27162

maropu commented:
retest this please

SparkQA commented:
**[Test build #116507 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116507/testReport)** for PR 27166 at commit [`6401164`](https://github.com/apache/spark/commit/640116443b362fd130f1da32c7e57365831e33fa).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21295/
Test PASSed.

SparkQA commented:
**[Test build #116507 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116507/testReport)** for PR 27166 at commit [`6401164`](https://github.com/apache/spark/commit/640116443b362fd130f1da32c7e57365831e33fa).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class FilterStatistics(`
  * `class EventLogFileCompactor(`
  * `case class CompactionResult(code: CompactionResultCode.Value, compactIndex: Option[Long])`
  * `case class AlterNamespaceSetOwner(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116507/
Test FAILed.

MaxGekk committed:
Merge remote-tracking branch 'remotes/origin/master' into dedup-appender-skeleton

SparkQA commented:
**[Test build #116539 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116539/testReport)** for PR 27166 at commit [`5ede347`](https://github.com/apache/spark/commit/5ede347f093e5e1b606f448b1c01578066bce75a).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21325/
Test PASSed.

SparkQA commented:
**[Test build #116539 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116539/testReport)** for PR 27166 at commit [`5ede347`](https://github.com/apache/spark/commit/5ede347f093e5e1b606f448b1c01578066bce75a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116539/
Test FAILed.

dongjoon-hyun commented:
Retest this please.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21328/
Test PASSed.

SparkQA commented:
**[Test build #116542 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116542/testReport)** for PR 27166 at commit [`5ede347`](https://github.com/apache/spark/commit/5ede347f093e5e1b606f448b1c01578066bce75a).

SparkQA commented:
**[Test build #116542 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116542/testReport)** for PR 27166 at commit [`5ede347`](https://github.com/apache/spark/commit/5ede347f093e5e1b606f448b1c01578066bce75a).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116542/
Test FAILed.

MaxGekk commented:
Not sure how it is related to my changes:
```
======================================================================
ERROR: test_memory_limit (pyspark.tests.test_worker.WorkerMemoryTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jenkins/workspace/SparkPullRequestBuilder/python/pyspark/tests/test_worker.py", line 193, in test_memory_limit
    actual = rdd.map(lambda _: getrlimit()).collect()
  File "/home/jenkins/workspace/SparkPullRequestBuilder/python/pyspark/rdd.py", line 889, in collect
    sock_info = self.ctx._jvm.PythonRDD.collectAndServe(self._jrdd.rdd())
  File "/home/jenkins/workspace/SparkPullRequestBuilder/python/lib/py4j-0.10.8.1-src.zip/py4j/java_gateway.py", line 1286, in __call__
    answer, self.gateway_client, self.target_id, self.name)
  File "/home/jenkins/workspace/SparkPullRequestBuilder/python/lib/py4j-0.10.8.1-src.zip/py4j/protocol.py", line 328, in get_return_value
    format(target_id, ".", name), value)
Py4JJavaError: An error occurred while calling z:org.apache.spark.api.python.PythonRDD.collectAndServe.
: org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 0.0 failed 1 times, most recent failure: Lost task 0.0 in stage 0.0 (TID 0, amp-jenkins-worker-05.amp, executor driver): org.apache.spark.SparkException: Python worker exited unexpectedly (crashed)
```

MaxGekk commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/OptimizerLoggingSuite.scala:
```diff
@@ -39,45 +36,25 @@ class OptimizerLoggingSuite extends PlanTest {
-    override def requiresLayout(): Boolean = false
-  }
-
-  private def withLogLevelAndAppender(level: Level, appender: Appender)(f: => Unit): Unit = {
```
This is removed because the method is not used in the test suite.

MaxGekk commented:
The difference between the master and this PR is some tests could put more log event in log appender like https://github.com/apache/spark/pull/27166/files#diff-5a20f7138b31f3c344a4af7ab6f426daL45 but log appender should be removed at the end of the test. The problem could happen if a log appender is not removed and continue gather other log events.

MaxGekk committed:
Limit the number of event stored in LogAppender

MaxGekk committed:
> -> >=

SparkQA commented:
**[Test build #116549 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116549/testReport)** for PR 27166 at commit [`a054624`](https://github.com/apache/spark/commit/a054624a70e57e052ece2725e9f8b7cc423cc4a5).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21335/
Test PASSed.

SparkQA commented:
**[Test build #116549 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116549/testReport)** for PR 27166 at commit [`a054624`](https://github.com/apache/spark/commit/a054624a70e57e052ece2725e9f8b7cc423cc4a5).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116549/
Test FAILed.

MaxGekk committed:
Revert "Move LogAppender to SparkFunSuite"

This reverts commit e17793c4b29eece61db96d849c6a170fa4b33f5f.

HyukjinKwon commented:
@MaxGekk, I can't take a close look right now but it might be related to the Jenkins machine itself, and it happens consistently in some machines only. This can be checked via Jenkins log. I plan to check it tomorrow but it might be a good hint for you to investigate further.

SparkQA commented:
**[Test build #116553 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116553/testReport)** for PR 27166 at commit [`83b4b70`](https://github.com/apache/spark/commit/83b4b7061ee3647a5aaabd1767a2bb2a811c3aaa).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21339/
Test PASSed.

SparkQA commented:
**[Test build #116553 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116553/testReport)** for PR 27166 at commit [`83b4b70`](https://github.com/apache/spark/commit/83b4b7061ee3647a5aaabd1767a2bb2a811c3aaa).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  class MockAppender extends AppenderSkeleton `
  * `    class MockAppender extends AppenderSkeleton `
  * `  class MockAppender extends AppenderSkeleton `
  * `  class MockAppender extends AppenderSkeleton `
  * `    class TestAppender extends AppenderSkeleton `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116553/
Test FAILed.

MaxGekk commented:
jenkins, retest this, please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21343/
Test PASSed.

SparkQA commented:
**[Test build #116557 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116557/testReport)** for PR 27166 at commit [`83b4b70`](https://github.com/apache/spark/commit/83b4b7061ee3647a5aaabd1767a2bb2a811c3aaa).

SparkQA commented:
**[Test build #116557 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116557/testReport)** for PR 27166 at commit [`83b4b70`](https://github.com/apache/spark/commit/83b4b7061ee3647a5aaabd1767a2bb2a811c3aaa).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  class MockAppender extends AppenderSkeleton `
  * `    class MockAppender extends AppenderSkeleton `
  * `  class MockAppender extends AppenderSkeleton `
  * `  class MockAppender extends AppenderSkeleton `
  * `    class TestAppender extends AppenderSkeleton `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116557/
Test PASSed.

MaxGekk committed:
Revert "Revert "Move LogAppender to SparkFunSuite""

This reverts commit 83b4b7061ee3647a5aaabd1767a2bb2a811c3aaa.

SparkQA commented:
**[Test build #116565 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116565/testReport)** for PR 27166 at commit [`5fc87c4`](https://github.com/apache/spark/commit/5fc87c421ca8750e3d187a8c294634565eaef608).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21351/
Test PASSed.

SparkQA commented:
**[Test build #116565 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116565/testReport)** for PR 27166 at commit [`5fc87c4`](https://github.com/apache/spark/commit/5fc87c421ca8750e3d187a8c294634565eaef608).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  class LogAppender(maxEvents: Int = 100) extends AppenderSkeleton `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116565/
Test FAILed.

MaxGekk commented:
I reverted all the changes of this PR, and python tests passed (https://github.com/apache/spark/pull/27166#issuecomment-573439173), but when I reverted everything back - the test failed again (https://github.com/apache/spark/pull/27166#issuecomment-573451307). At the moment, I cannot figure out how my changes impact on the python test.

HyukjinKwon commented:
retest this please

HyukjinKwon commented:
Should be fixed as of https://github.com/apache/spark/commit/0823aec4630e70323e66bea243871aaab761d9ca ...

SparkQA commented:
**[Test build #116628 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116628/testReport)** for PR 27166 at commit [`5fc87c4`](https://github.com/apache/spark/commit/5fc87c421ca8750e3d187a8c294634565eaef608).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21409/
Test PASSed.

SparkQA commented:
**[Test build #116628 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116628/testReport)** for PR 27166 at commit [`5fc87c4`](https://github.com/apache/spark/commit/5fc87c421ca8750e3d187a8c294634565eaef608).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  class LogAppender(maxEvents: Int = 100) extends AppenderSkeleton `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116628/
Test PASSed.

MaxGekk commented:
@HyukjinKwon 's fix helped, @maropu this can be merged.

maropu commented:
yea, I will, thanks!

maropu commented:
Merged to master.
